### PR TITLE
Added annotate and classify functions to DPP

### PR DESCRIPTION
**Pull request by hsotoparada**
Hey @yetinam
This is a new PR to include the classify/annotate functions for the classes DPPDetector and DPPPicker.

In DPPDetector class, I have added the classify_aggregate method, which works similarly as in GPD, with an additional method _select_picks. This additional method includes optional conditions that are applied to remove presumed false positive predicted by DPPDetector (e.g., S picks for which there is no earlier P pick predicted or possible duplicated S phases) as well as to discriminate between P and S phases predicted close in time (with overlapping or non-overlapping probability functions).

In DPPPicker class, I have overwritten the classify/annotate functions as described below, trying to follow the structure and functionality of equivalent methods in the WaveformModel class. 

In the overwritten annotate function:
The method stream_to_arrays converts streams (associated to each preliminary pick coming from a detector, e.g., DPPDetector) into a list of start times and numpy arrays of trimmed streams within the picking window. This picking window would be defined by the optimized hyperparameters in DPP.

The method _annotate_window (the equivalent to _annotate_point for picking) runs the picker on the picking windows, returning a list of arrays containing the results from the picking run, that is, the probabilities resulting from applying the Monte Carlo Dropout MCD iteration (which would be the characteristic function/probability function over time within the picking windows).

The method _predictions_to_stream returns streams containing the results from picking.

And in the overwritten classify_aggregate function:
The method picks_from_annotations returns a pick object for each annotation (each stream with results from picking).

I am still working on this implementation, but it'd be nice to know if it seems reasonable to you until now (especially for the picker method). Also just let me know about any suggestions that you may have to improve it.
For example, I think that instead of overwriting some of these methods, their functionality could be integrated into the methods already defined in WaveformModel. This could also work for the overwritten Pick class, which is just an extended version of the Pick class in seisbench.util.annotations.Pick.

For further testing this implementation, I would like now to see how to integrate the trained weights into seisbench.
From what I've seen a json file is used for that, in which I think it could make sense to additionally include the optimized hyperparameters in the case of DPP.
But I don't know how I could create this file for DPP. Could you guide me on this or indicate me the best way to do this?

This branch is not ready for merging yet.

### CODE REVIEW

File: seisbench/models/dpp.py [hunk 1]
```diff
@@ -223,3 +526,525 @@ def forward(self, x):
...
+
+                # Write to output stream.
+                # Returns streams containing results from MCD iteration.
+                output[-1] += self._predictions_to_stream(
+                    pred_rates, pred_times, preds, pick_in, trace
+                )
+
+        return output
+
+    def _predictions_to_stream(self, pred_rates, pred_times, preds, pick_in, trace):
```
[the remark this answers is not in the document]
Comment (hsotoparada) [reply seisbench/models/dpp.py:631]: I removed `_predictions_to_stream`. Just for reference, the intended purpose of this function was:
"Return stream containing results from applying MCD (MCD samples), which populate the output of `annotate`".

File: seisbench/models/dpp.py [hunk 2]
```diff
@@ -223,3 +526,525 @@ def forward(self, x):
...
+                            "location": trace.stats.location,
+                            "channel": f"{self.__class__.__name__}_{label}",
+                        },
+                    )
+                )
+
+        return output
+
+    # runs Monte Carlo Dropout (MCD) on picking windows, returning mc_pred array (as preds) containing results from MCD iteration
+    def _annotate_window(self, times, data, pick_in, argdict):
```
[the remark this answers is not in the document]
Comment (hsotoparada) [reply seisbench/models/dpp.py:665]: I removed `_annotate_window`.  Just for reference, the intended purpose of this function was:
"Run MCD on picking windows (previously cut out from data stream), returning a list of arrays each containing MCD samples for a predicted pick".
It may be helpful for implementing it into `_annotate_array` in `base.py`.

File: seisbench/models/dpp.py [hunk 3]
```diff
@@ -223,3 +526,525 @@ def forward(self, x):
...
+            pred_rates.append(argdict["sampling_rate"])
+            # mc_pred = np.array(mc_pred)[:,0,:,:] # mc_pred.shape = (mc_iter, block.shape[1], 1)
+            mc_pred = np.array(mc_pred)[
+                :, 0, :, 0
+            ].T  # mc_pred.shape = (block.shape[1], mc_iter)
+            full_preds.append(mc_pred)
+
+        return pred_times, pred_rates, full_preds
+
+    def stream_to_arrays(
```
[the remark this answers is not in the document]
Comment (hsotoparada) [reply seisbench/models/dpp.py:714]: Okey, I removed `stream_to_arrays`. Just for reference, the intended purpose of this function was:
"From data stream, cut out picking windows around preliminary picks obtained from `DPPDetector`"

File: seisbench/models/dpp.py [hunk 4]
```diff
@@ -223,3 +526,525 @@ def forward(self, x):
...
+                for trace in traces:
+                    if t1 < trace.stats.endtime:
+                        start_sorted.put((t1, seqnum, trace.slice(starttime=t1)))
+                        seqnum += 1
+
+        return output_times, output_data
+
+    def classify_aggregate(self, annotations, picks_in, argdict):
+        """
+        Converts the annotations to discrete picks using :py:func:`~seisbench.models.base.WaveformModel.picks_from_annotations`.
```
[the remark this answers is not in the document]
Comment (hsotoparada) [reply seisbench/models/dpp.py:891]: Docstring has been adapted, now it looks as shown below.
The `[phase]_threshold` key is still used, but with slightly different meaning that in detection models, since here it indicates the threshold (by default 0.5) used in the overwritten `picks_from_annotations` function to identify the pick onset from the MCD samples.

```
def classify_aggregate(self, annotations, argdict):                         
    """                                                                     
    Converts the annotations (MCD samples) to discrete picks using :py:func:`picks_from_annotations`.
    Thresholds for pick onsets (transition from noise to signal) are derived from the argdict at keys "[phase]_threshold".
                                                                            
    :param annotations: See description in superclass                       
    :param argdict: See description in superclass                           
    :return: List of picks                                                  
    """       
```

File: seisbench/models/dpp.py [hunk 5]
```diff
@@ -223,3 +526,525 @@ def forward(self, x):
...
+                argdict,
+                phase,
+            )
+
+        return sorted(picks)
+
+    @staticmethod
+    def picks_from_annotations(annotation, argdict, phase):
+        """
+        Converts the annotations streams for a single phase to discrete picks using a classical trigger on/off.
```
[the remark this answers is not in the document]
Comment (hsotoparada) [reply seisbench/models/dpp.py:915]: Docstring has been adapted, now it looks as shown below.

```
@staticmethod                                                               
def picks_from_annotations(annotation, threshold, phase):                   
    """                                                                     
    Converts the annotations (MCD samples) for a single pick to discrete picks including pick uncertainties.
    A predefined probability threshold is used to identify the pick onset as the timestep where the transition from noise to 
    signal occurs.
    Picks are represented by :py:class:`~seisbench.util.annotations.Pick` objects.
    The pick start_time and end_time represent the pick uncertainty, as obtained from the standard deviation of the MCD 
    inferences applied for the pick.
                                                                            
    :param annotations: Stream of annotations (MCD samples)                 
    :param threshold: Probability threshold for the pick onset              
    :param phase: Phase to label, only relevant for output phase labelling  
    :return: List of picks                                                  
    """                      
```

File: seisbench/models/dpp.py [hunk 6]
```diff
@@ -223,3 +526,525 @@ def forward(self, x):
...
+        Converts the annotations to discrete picks using :py:func:`~seisbench.models.base.WaveformModel.picks_from_annotations`.
+        Trigger onset thresholds for picks are derived from the argdict at keys "[phase]_threshold".
+
+        :param annotations: See description in superclass
+        :param argdict: See description in superclass
+        :return: List of picks
+        """
+        picks = []
+        for annotation in annotations:
+            phase = annotation[0].stats.channel.split("_")[-1][0]
```
[the remark this answers is not in the document]
Comment (hsotoparada) [reply seisbench/models/dpp.py:900]: Exactly, my idea was that those annotations (MCD samples) were returned as a list of streams by the overwritten `annotate` function. It was just the way I thought this could be implemented following the existent logic for detection.
But I can see this would be inconsistent with the functionality intended for other models, so I'm okay with your suggestions.
However, `classify_aggregate` would still need to receive as input annotations from `annotate` which should contain the MCD samples for predicted picks.

File: seisbench/models/dpp.py [hunk 7]
```diff
@@ -223,3 +526,525 @@ def forward(self, x):
...
+            peak_uncertainty=[terr_pre, terr_pos],
+            peak_class=pick_class,
+            phase=phase,
+        )
+
+        return pick
+
+
+# ~seisbench.util.annotations.Pick adapted to DPPPicker
+class Pick:
```
[the remark this answers is not in the document]
Comment (hsotoparada) [reply seisbench/models/dpp.py:1004]: Done as suggested, so I removed the overwritten `Pick` class.
In `peak_value=pick_class`, higher values actually come from higher uncertainties, that indicate lower quality (best pick class is 0). In any case, I imagine this misuse would still be acceptable.

File: seisbench/models/dpp.py [hunk 8]
```diff
@@ -137,10 +142,308 @@ def forward(self, x):
...
+    # TODO: update
+    def __init__(
+        self,
+        mode,
+        in_channels=3,
+        phases=None,
+        sampling_rate=100,
+        **kwargs,
+    ):
+        super().__init__(
```
[the remark this answers is not in the document]
Comment (hsotoparada) [reply seisbench/models/dpp.py:441]: Done

File: seisbench/models/dpp.py [hunk 9]
```diff
@@ -137,10 +142,308 @@ def forward(self, x):
...
+            sampling_rate=sampling_rate,
+            labels=phases,
```
[the remark this answers is not in the document]
Comment (hsotoparada) [reply seisbench/models/dpp.py:443]: Done.

File: seisbench/models/dpp.py [hunk 10]
```diff
@@ -137,10 +142,308 @@ def forward(self, x):
...
+                argdict.get(f"{phase}_threshold", 0.9),
+                phase,
+            )
+
+        picks_out = _select_picks(self, annotations, picks, argdict)
+
+        return sorted(picks_out)
+
+    def _select_picks(self, annotations, picks, argdict):
+        """
```
[the remark this answers is not in the document]
Comment (hsotoparada) [reply seisbench/models/dpp.py:171]: I updated the docstrings of the `_select_picks` function as well as the individual functions that apply the optional checks.
Docstring of `_select_picks` now includes specific checks used from `argdict`.
Docstring of each individual function describes the specific check applied and which parameters from `argdict` are used for that check.
